webmin/os-eol-lib.pl

```diff
@@ -0,0 +1,228 @@
+                die("Could not fetch OS EOL data: " . $ferror);
+                }
+        my $fdata_json;
+        eval { $fdata_json = &convert_from_json($fdata); };
+        if ($@) {
+                die("Could not parse fetched OS EOL data: $@");
+                }
+        # Add OS
+        $fdata_json = [ map { $_->{'_os'} = $os; $_ } @$fdata_json ];
+        # Fix (blessed) LTS key
+        $fdata_json = [ map { $_->{'lts'} = $_->{'lts'} ? 1 : 0; $_ } @$fdata_json ];
+        push(@eol_oses_data, @$fdata_json);
+        }
+&write_file_contents($eol_cache_file, &convert_to_json(\@eol_oses_data));
+}
+
+# eol_get_os_data()
+# Returns EOL data for the current OS.
+# Returns undef if OS is not supported.
+sub eol_get_os_data
+{
+my $os = &eol_get_os();
+return undef if (!$os);
+my $eol_file = shift;
+$eol_file ||= "$root_directory/os_eol.json";
+if (!-r $eol_file) {
+        &error_stderr("Could not read OS EOL data file: $eol_file");
+        return undef;
+        }
+my $eol_data = &read_file_contents($eol_file);
+my $eol_json;
+eval { $eol_json = &convert_from_json($eol_data); };
+if ($@) {
+        &error_stderr("Could not parse OS EOL data: $@");
+        return undef;
+        }
+if (ref($eol_json) eq 'ARRAY' && @$eol_json) {
+        my $os_version = $gconfig{'real_os_version'};
+        # Extract the major and minor versions
+        $os_version =~ m/^(?:stable\/)?(?<major>\d+)(?:\.(?<minor>\d+))?/;
+        $os_version = $+{minor} ? "$+{major}.$+{minor}" : $+{major};
+        return undef if (!$+{major});
+        # Minor versions in cycle are allowed only for Ubuntu
+        $os_version = $+{major} if ($os !~ /ubuntu/);
+        my ($eol_json_this_os) =
+                grep { $_->{'_os'} eq $os &&
+                       $_->{'cycle'} eq $os_version } @$eol_json;
+        $eol_json_this_os->{'_os_name'} = $gconfig{'real_os_type'};
+        $eol_json_this_os->{'_os_version'} = $os_version;
+        # Convert EOL date to a timestamp
+        my ($year, $month, $day) = split('-', $eol_json_this_os->{'eol'});
+        $eol_json_this_os->{'_eol_timestamp'} = timelocal(0, 0, 0, $day, $month - 1, $year);
+        # Convert EOL extendend date to a timestamp
+        if ($eol_json_this_os->{'extendedSupport'}) {
+                my ($year, $month, $day) = split('-', $eol_json_this_os->{'extendedSupport'});
+                $eol_json_this_os->{'_ext_eol_timestamp'} =
+                        timelocal(0, 0, 0, $day, $month - 1, $year);
+                }
+        return $eol_json_this_os if ($eol_json_this_os);
+        }
+return undef;
+}
+
+# eol_populate_dates(&eol_data)
+# Updates given EOL hash reference with
+# human readable date and the time ago
+sub eol_populate_dates
+{
+my ($eol_data) = @_;
+if (!$eol_data->{'_eol_timestamp'}) {
+        &error_stderr("The provided data is not a valid EOL data hash reference");
+        return undef;
+        }
+my $eol_date = &make_date($eol_data->{'_eol_timestamp'}, { '_' => 1 });
```

Comment:
Just curious, what does the `_` option mean?

Reply:
This is a pseudo option to tell `make_date` return an object.